src/protocol/basics/reshare.rs

```diff
@@ -30,36 +30,39 @@ use futures::future::try_join;
 ///    `to_helper`       = (`rand_left`, `rand_right`)     = (r0, r1)
 ///    `to_helper.right` = (`rand_right`, part1 + part2) = (r0, part1 + part2)
 #[async_trait]
-pub trait Reshare<V: SharedValue> {
-    type Share: SecretSharing<V>;
-
-    async fn reshare(
-        self,
-        input: &Self::Share,
-        record: RecordId,
+pub trait Reshare<C: Context>: Sized {
+    async fn reshare<'fut>(
+        ctx: C,
+        record_id: RecordId,
+        input: &Self,
```

Comment:
Any reason that this isn't `&self`?  I don't see why not.

Reply:
We could do it that way, and it would save needing to call the methods as `Type::method`. (I don't find `T::method` too bad; in other places it is `Replicated::method` or `MaliciousReplicated::method` which is worse.)

I kind of like having the context be the first argument (and that's what we do in a bunch of existing places), but there's no reason it has to be. But, for multiplies, we have to take a 2nd operand:
* `Replicated::multiply(ctx, record, a, b)`
* `a.multiply(ctx, record, b)`
* `a.multiply(b, ctx, record)`
* `(a, b).multiply(ctx, record)` (I think this goes in the "too clever for it's own good" department.)

I would probably pick the 3rd over the 2nd (keep the operands closer together), but that introduces an inconsistency if we otherwise put `ctx` first.

Reply:
The third is better, yeah.  I am not a huge fan of having to bring in the trait (with `use`) and then use `Struct::method` to invoke what is better conceived of as a function on the operands.  If only because it is more verbose than necessary and finding the right type to use as a base is awkward.

Reply:
Oh, I should say, I'm going to merge this now; do that part later :)